### PR TITLE
Re-organize packages: Build, design and runtime.

### PR DESCRIPTION
**Pull request by peterp**
This PR is for mainly for the maintainers of RedwoodJS and wont affect users.

RedwoodJS is a framework that relies upon a bunch of 3rd party dependencies. As of June 2021 **contributors install about 900 MB of node_modules.** That's a lot of code!

image

It doesn't feel nice when there are so many packages! It's overwhelming to understand where all of them are used, what to test when you upgrade a package, and more packages just lead to slower installation times... It's seems out of control.

On the other hand this is how modern software is built. Dependencies allow us to focus on the problem we're trying to solve, and it allows us to move way faster!

I think we can ease some of this pain through organization. But, how? It makes sense to focus on when the packages are used in the dev to production life-cycle. I have broken this life-cycle down into three phases:
- **Build time**
Forms the foundation, used to convert the code in repository to something that's deployable to production: Transpilation, code generation, and bundling.
- **Design time**
Used during development: Testing, linting, component playgrounds (storybook), and plugins that support development.
- **Runtime**
Required by the code that's deployed to staging/ production: HTTP servers, React, Apollo GraphQL, etc.

image

These phases are implied but never communicated or organized, this PR starts that work of enforcing those boundaries.

Note: These phases aren't 100% technically perfect, but the metaphor is helpful for organization purposes.

- [x] Move build time packages into `@redwoodjs/core`
- [x] Move design time packages into `@redwoodjs/testing`
- [x] Add `core` and `testing` to CLI. 
- [x] Install `@redwoodjs/cli` instead of `@redwoodjs/core` in template.
- [x] Re-organize configurations: Put testing configs into testing. Deprecate `getConfig` method.
- [x] Move `StorybookProvider` into testing. 
- [x] The framework build-time packages are removed from the workspace, and placed into `tasks/framework-tools`

### CODE REVIEW

File: packages/cli/package.json [hunk 1]
```diff
@@ -13,11 +13,16 @@
     "dist"
   ],
   "dependencies": {
-    "@prisma/sdk": "2.24.01",
     "@redwoodjs/api-server": "^0.33.0",
     "@redwoodjs/internal": "^0.33.0",
     "@redwoodjs/prerender": "^0.33.0",
     "@redwoodjs/structure": "^0.33.0",
+    "@redwoodjs/cli": "^0.33.0",
```
Comment (peterp) [inline packages/cli/package.json:20]:
```suggestion
```

**Comment by peterp** [comment]
@thedavidprice It's ready for review and test. The text is just for context if anyone wanders into this.

File: package.json [hunk 2]
```diff
@@ -60,5 +19,21 @@
     "vscode-languageserver-types": "3.15.1",
     "vscode-languageserver": "6.1.1",
     "vscode-languageserver-textdocument": "1.0.1"
+  },
+  "scripts": {
+    "framework": "yarn --cwd ./tasks/framework-tools",
```
Comment (thedavidprice) [inline package.json:24]: this is really nice 🤩

File: packages/cli/package.json [hunk 3]
```diff
@@ -13,11 +13,15 @@
...
   ],
   "dependencies": {
-    "@prisma/sdk": "2.24.1",
     "@redwoodjs/api-server": "^0.33.2",
     "@redwoodjs/internal": "^0.33.2",
     "@redwoodjs/prerender": "^0.33.2",
     "@redwoodjs/structure": "^0.33.2",
+    "@redwoodjs/dev-server": "^0.33.0",
+    "@redwoodjs/eslint-config": "^0.33.0",
+    "@redwoodjs/core": "^0.33.0",
```
Comment (thedavidprice) [inline packages/cli/package.json:22]:
```suggestion
    "@redwoodjs/dev-server": "^0.33.2",
    "@redwoodjs/eslint-config": "^0.33.2",
    "@redwoodjs/core": "^0.33.2",
```

File: packages/cli/package.json [hunk 4]
```diff
@@ -13,11 +13,15 @@
...
+    "prisma": "2.24.1",
+    "@prisma/sdk": "2.24.01",
```
Comment (thedavidprice) [inline packages/cli/package.json:24]:
```suggestion
    "@prisma/sdk": "2.24.1",
```

File: packages/cli/src/index.js [hunk 5]
```diff
@@ -12,6 +12,10 @@ config({
   defaults: path.join(getPaths().base, '.env.defaults'),
 })
 
+// TODO:
+// Add structure and rwjsPaths to global context.
+// Add 'cwd' command.
+
```
Comment (thedavidprice) [inline packages/cli/src/index.js:18]: reminder if for this PR

File: packages/core/package.json [hunk 6]
```diff
@@ -13,70 +14,59 @@
...
+    "url-loader": "4.1.1",
+    "terser-webpack-plugin": "4.2.3",
+    "@pmmmwh/react-refresh-webpack-plugin": "0.4.3",
+    "mini-css-extract-plugin": "1.4.1",
+    "null-loader": "4.0.1",
+    "lodash.escaperegexp": "4.1.2",
+    "esbuild-loader": "2.13.1",
+    "x----x----x": "^0 RedwoodJS",
+    "@redwoodjs/internal": "^0.33.0",
+    "@redwoodjs/testing": "^0.33.0"
```
Comment (thedavidprice) [inline packages/core/package.json:69]:
```suggestion
    "@redwoodjs/internal": "^0.33.2",
    "@redwoodjs/testing": "^0.33.2"
```

File: packages/cli/package.json (hunk 3 above)
Comment (thedavidprice) [reply packages/cli/package.json:22]: Also, I just noticed Redwood packages regained the `^`, which I assume is from Lerna.

Feels like we should pin as well. (If so do you know if/how to manage that for Lerna?) Either way will defer to your judgement.

**Review by thedavidprice — COMMENTED** [review]
@peterp Way to go on this. I'm already feeling better about my Webpack and Jest PRs.

I did my best to test locally in a project — upgraded to canary and then used rwt copy:watch. I tested commands and functionality that seemed most likely to be affected by these changes (and isn’t necessarily included in CI). Wasn't much. All seem to be working!

There are some misc package.json edits regarding versions to be fixed. I left comments for each.

## @redwoodjs/core --> cli
@peterp I didn’t see an update to crwa/template/package.json swapping core for cli package. That is required for this change, correct?

Related:
- [ ] please also update this line in `yarn rw info`, which used the @redwoodjs package in the root package.json for version: https://github.com/redwoodjs/redwood/blob/1a03b9797784bb372c60858fa603668ab3d603a2/packages/cli/src/commands/info.js#L23

## rwt link and build:test-project
@dac09 I couldn't get `rwt link` to work, which was unexpected. You might want to take a look at this PR to see if there's anything obvious. If we merge and it breaks, a lot of folks are going to come asking for help (me included)!

**Comment by thedavidprice** [comment]
## Breaking
The following files require projects to update:
- /api/jest.config.js
- web/jest.config.js
- package.json

**Comment by thedavidprice** [comment]
@peterp lingering thoughts about moving core into cli --> the decision makes sense but it feels a bit non-intuitive from a naming perspective. I started wondering if we could make a transition like prisma that moved @prisma/cli to just prisma, which would mean we release a 'redwoodjs' that included cli and core (the package isn't available, btw). (A part of their rationale was that it makes it easier for people to run CLI commands via `npx prisma ...`, which seems useful for us as well.)

I agree that fewer packages in the template is better and I know there's a longer roadmap at play. So I'm not sure if there's anything actionable here. Perhaps we choose a new package name that represents something foundational, like 'main'. 

If something strikes you, go with it. Otherwise, consider this fwiw...

File: packages/cli/src/index.js [hunk 7]
```diff
@@ -12,6 +12,10 @@ config({
   defaults: path.join(getPaths().base, '.env.defaults'),
 })
 
+// TODO:
+// Add structure and rwjsPaths to global context.
+// Add 'cwd' command.
```
Comment (peterp) [inline packages/cli/src/index.js:17]:
```suggestion
```

File: packages/cli/package.json (hunk 3 above)
Comment (peterp) [reply packages/cli/package.json:22]: Good point - I'll look into lerna's docs to figure this out!

**Comment by peterp** [comment]
@dac09 I see `yarn build:link` was indeed broken because the `packlist` was in the wrong place.

I think I've noticed what may be a bug in `yarn build:link`, am I correct to say that it assumes you've run `yarn rwt link` before you've run `yarn build:link`?

**Comment by peterp** [comment]
@dac09 I've made `yarn rwt link` copy the packages into the same `node_modules` folder as the e2e task: 'node/modules/.redwood-linked-packages'.

This has the advantage that the folder is right at the top of the directory structure, and having them the same means that we don't have e2e redwood packages and link redwood packages.

**Comment by dac09** [comment]
Hey @peterp I don't actually fully follow whats going on in this PR (I totally dig the high level idea though).

- But to help with your reasoning - `build:link` can run independantly of rwt link, but rwt link does internally call build:link ofcourse.
- Packlist should be a internal dependency of redwood, its not needed for the cli. Packlist is only used if you're trying to use your local version of redwood
- I'm not sure whats happening the with cli package, the function `updateProjectWithResolutions` in redwood-tools might be worth checking - It uses the cli package's version to set resolutions for all the packages. Maybe we could change this to match one of the newer packages instead of the cli?

**Comment by dac09** [comment]
@peterp are you sure about it being in a hidden folder (i.e. with a dot in front?). My reasoning for not doing this was so that incase something goes wrong in link, or you forgot to unlink, you'd be able to see the folder in your node_modules. Then again, who even looks inside that folder (apart from us)

**Comment by peterp** [comment]
> But to help with your reasoning - build:link can run independantly of rwt link, but rwt link does internally call build:link ofcourse.

Ok, so the flaw that I'm seeing is that `rwt link` creates a folder in node_modules (`node_modules/.redwood-linked-packages`), adds that folder to the workspace, and then run `yarn install`. This installs the new dependencies into a RedwoodJS project. (This "workspace hack" is what enables links behaviour)

`yarn build:link` skips the step above and instead copies the files into `node_modules/@redwood/*` so whilst this is OK for updating the code, it would not install new packages (which is what we're trying to solve with link, right?).

So, the potential flaw in having these two divergent behaviour for the "same concept" may be confusing to contributors, in that if I have a project and only run `yarn build:link --dest path/to/project` it will work very differently from running `yarn rwt link`.

I hope that makes sense?

**Comment by dac09** [comment]
> So, the potential flaw in having these two divergent behaviour for the "same concept" may be confusing to contributors, in that if I have a project and only run yarn build:link --dest path/to/project it will work very differently from running yarn rwt link.

I believe build-link just copies it to the folder you specify, not to @redwood/* automatically. The scope params with @redwood you see there are just to tell lerna to only build specific packages.  The behaviour I believe is the same, only difference is `rwt link` just does the extra steps to setup _your project_. Build link is only concerned with the framework, and copies it to wherever you tell it to.

That being said, I don't expect anyone actually uses build-link without rwt link!

**Comment by peterp** [comment]
> @peterp are you sure about it being in a hidden folder (i.e. with a dot in front?). My reasoning for not doing this was so that incase something goes wrong in link, or you forgot to unlink, you'd be able to see the folder in your node_modules.

So, in my case I actually noticed I only ever really look into `node_modules` via the terminal or via vscode, and in both cases it's visible and right at the top.

I'm also wondering if npm/ yarn ignores those, since it's what they use for `.bin` 🤷

**Comment by peterp** [comment]
> I believe build-link just copies it to the folder you specify, not to @redwood/* automatically. 

Ah ok, do you think there's value in converging these two into a single place? Maybe we can remove the yarn command from the workspace too.

**Comment by peterp** [comment]
@thedavidprice I really like the idea of just using a top-level `redwoodjs` package. Nice!

**Comment by dac09** [comment]
> Ah ok, do you think there's value in converging these two into a single place? Maybe we can remove the yarn command from the workspace too.

Yeah I think that could work! The only thing you need to watch out for is that: build-link (or equivalent) has to run in the scope of the package i.e. build-link needs to run individually for cli, internal, web..... which is why I was using the lerna command
This is because of `packlist` which will check what should be checked for each of those packages, and copies them across, not the toplevel one. (Does that make sense at all?)

**Comment by peterp** [comment]
> Yeah I think that could work! The only thing you need to watch out for is that: build-link (or equivalent) has to run in the scope of the package i.e. build-link needs to run individually for cli, internal, web..... which is why I was using the lerna command

Totes! I won't do any of this in this package, but I'll add a note for the subsequent tasks for package renaming stuff that's in phase two of this PR.

File: packages/cli/src/redwood-tools.js [hunk 8]
```diff
@@ -160,7 +160,7 @@ const rwtLink = async (yargs) => {
   const projectPackagesPath = path.join(
     getPaths().base,
     'node_modules',
-    '.redwood-linked-packages'
+    '_redwood-linked-packages'
```
Comment (dac09) [inline packages/cli/src/redwood-tools.js:163]: @peterp remember to upload `unlink` as well please!

**Comment by peterp** [comment]
@dac09 Totally makes sense - I'll make sure that it's working properly. We need tests for link!

**Comment by peterp** [comment]
@dac09 It's working as expected 👍

**Comment by thedavidprice** [comment]
> @thedavidprice I really like the idea of just using a top-level redwoodjs package. Nice!

Thoughts on what it could be?

**Comment by peterp** [comment]
Converting this to draft so that it doesn't accidentally get merged.

**Comment by thedavidprice** [comment]
@peterp could we hold on merging this until we can discuss Envelop PR and testing?

**Review by thedavidprice — CHANGES_REQUESTED** [review]
ping

**Review by thedavidprice — APPROVED** [review]

**Comment by thedavidprice** [comment]
## Note on /cli --> /core
> Install @redwoodjs/cli instead of @redwoodjs/core in template.

This was reverted in #2825

**Pull request merged**